tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,269 @@
+    """
+
+    def __init__(self, config, session, logger):
+        """
+        Initialize SplunkHecDelivery HEC sender.
+
+        :param config: sqs_splunk_lambda configuration
+        :type config: dict
+        :param session: boto3 AWS Session
+        :param logger: Logger object to write to
+        """
+        self.config = config
+        self.logger = logger
+        self.session = session
+
+    def get_splunk_payloads(self, msg, msg_timestamp):
+        """
+        Given a raw c7n message dict, generate a list of payloads to send to the
+        Splunk HEC.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :param msg_timestamp: int timestamp (milliseconds) when message was sent
+          to SQS, for use as event timestamp in Splunk.
+        :type msg_timestamp: int
+        :return: list of Splunk payload dicts
+        :rtype: list
+        """
+        payloads = []
+        events = self.get_splunk_events(msg)
+        indices = self._splunk_indices_for_message(msg)
+        for event in events:
+            for index in indices:
+                payloads.append({
+                    'time': msg_timestamp,
+                    'host': 'cloud-custodian',
+                    'source': '%s-cloud-custodian' % event.get(
+                        'account', 'unknown'
+                    ),
+                    'sourcetype': '_json',
+                    'index': index,
+                    'event': event
+                })
+        return payloads
+
+    def get_splunk_events(self, msg):
+        """
+        Given a raw c7n message dict, generate a list of per-resource event
+        dictionaries to send to Splunk as events.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :return: list of Splunk HEC Event dicts
+        :rtype: list
+        """
+        # if an event is present, add it to the log
+        if msg.get('event', None) is not None:
+            # add user to the log message, at the top level...
+            user = get_aws_username_from_event(self.logger, msg['event'])
+            if user is not None:
+                msg['event_triggering_user'] = user
+        # get a copy of the message with no resources
+        base_log = dict(msg)
+        base_log.pop('resources')
+        base_log = deepcopy(base_log)
+        # if configured, build and add actions list
+        if self.config.get('splunk_actions_list', False):
+            base_log['actions'] = []
+            for a in msg['policy']['actions']:
+                if isinstance(a, type({})):
+                    base_log['actions'].append(a['type'])
+                else:
+                    base_log['actions'].append(a)
+        # generate a separate Splunk message for each resource
+        logs = []
+        for res in msg['resources']:
+            x = dict(base_log)
+            if (
+                self.config.get('splunk_remove_metrics', False) and
```

Comment:
would this be more generically useful as removing all custodian annotations (ie any key starting with c7n)? or is it pretty specific to metrics?

Reply:
we should probably be operating on a shallow copy of the resource here as well to avoid the mods for other transports.

Reply:
I'm really not sure for other people. For us, we want to keep any other c7n keys, but we just want to dump the metrics because they can add quite a bit to the size of the log message and we've never had a need for them yet (if we did, we'd likely pull them from the S3 reporting).

Reply:
thats fine, let's just leave as metrics removal,  *but* wouldn't this functionality just fold into configuration of one of the splunk_remove_paths ? ie does it need a separate implementation.

Reply:
You're right. I've removed the ``splunk_remove_metrics`` functionality in favor of just using a ``splunk_remove_paths`` entry.